### PR TITLE
Careful handling when loading files

### PR DESCRIPTION
**Pull request by jentfoo**
This PR attempts to provide a common API so that the decision of when to follow symlinks is a conscious decision. Because Teleport (particularly the agent) runs in a privileged context, there is risk that following symlinks may allow information disclosure.

After review of the cases covered in this commit (and some additional cases where this API was not a natural fit), this does not appear to be a broad problem.  This PR however does fix the one known flaw described in the issue https://github.com/gravitational/teleport-private/issues/1009

In many cases where this API is being adopted, symlinks are still being allowed.  This is only after verifying there is no information disclosure when symlinks are followed.  Although symlinks are not likely useful in many of these cases, this choice to still allow symlinks is to maintain compatibility with potentially unknown use cases.  Future use of this API will almost certainly always specify a `false` for allowing symlinks.

### CODE REVIEW

File: lib/cgroup/cgroup.go
Comment (zmb3) [inline lib/cgroup/cgroup.go:175]: It's not clear at the call-site what the extra boolean is for. What do you think about a separate func for when we don't want to follow symlinks? Something like `OpenFileNoSymlinks(path)` (then we could just use the standard `os.Open` when we want the default behavior.

Comment (jentfoo) [reply lib/cgroup/cgroup.go:175]: I suppose we could as long as the symlink is verbose in naming `OpenFileAllowingSymlinks`.  The goal in having it be a parameter is that the developer must make a choice in what parameter is provided.

Let me know what you think would be less likely to have a copy paste error, I think either could be fine as long as the name is verbose.

Comment (zmb3) [reply lib/cgroup/cgroup.go:175]: > The goal in having it be a parameter is that the developer must make a choice in what parameter is provided.

Well, the developer doesn't have to make a choice if they don't know that we've created our own set of utilities that replace what the standard library offers. Most Go devs are not going to reach for these by default (especially because it's buried in a `utils` package with other misc and unrelated code).

I think it'd be better to use the standard library function when you want the default behavior of the standard library, and to use our overrides when you need something different.

Comment (jentfoo) [reply lib/cgroup/cgroup.go:175]: The goal is that "need something different" should be our default.  In general we should not allow the following of symlinks.  Do you have any thoughts on how we can make using this pattern a more default behavior?

Comment (codingllama) [reply lib/cgroup/cgroup.go:175]: +1 for "OpenFileNoSymlink" (or some other variation) instead of OpenFile(string, bool). Because OpenFile already exists is not the same as os.Open, I suppose we ought to keep both.

A way to get this through the codebase is to replace all usages, as at least when doing maintenance it should draw attention. Other than that, we could try and put together a linter of sorts.

File: lib/utils/fs.go
Comment (timothyb89) [inline lib/utils/fs.go:112]: A couple notes from things I learned trying to do this for `tbot`:
1. This check needs to use `os.Lstat()` to ever return `true`, since `os.Stat()` follows symlinks. We're just reading the destination file info here, which can never be a symlink, even if broken.
2. This one depends on our threat model, but even `os.Lstat()` follows symlinks, if they occur in any preceding path component. It's also a bit hairy to recursively check each path component since there could be cycles and other shenanigans.
   * This caveat applies to any solutions claiming to prevent symlink attacks using some variation of `O_NOFOLLOW`, since that flag only ever applies to the final pathname component.

We use [`openat2()` with `RESOLVE_NO_SYMLINKS`](https://man7.org/linux/man-pages/man2/openat2.2.html) in tbot to prevent this sort of file redirection attack: https://github.com/gravitational/teleport/blob/master/lib/tbot/botfs/fs_linux.go#L62

Unfortunately this comes with some pretty big caveats: `openat2()` isn't supported on older kernels, and is very frequently disabled on container engines and various enterprise Linuxes. And it's Linux-only :/. It's okay as an incremental security improvement where available, though - tbot checks for `openat2()` availability and falls back to insecure writes.

**Review by timothyb89 — CHANGES_REQUESTED** [review]

File: lib/cgroup/cgroup.go
Comment (jentfoo) [reply lib/cgroup/cgroup.go]: Updated the function naming.  I don't quite want to remove the ability to use `os.OpenFile` yet.  I am hopeful that this can still become the default when you need to open a read only file.

That said, we do need to keep in mind symlinks when writing too

File: lib/utils/fs.go
Comment (timothyb89) [reply lib/utils/fs.go:123]: It also looks like we evaluate symlinks ahead of time in `NormalizePath()`, so we probably need to be careful with its symlink-following behavior too :slightly_frowning_face:

Comment (jentfoo) [reply lib/utils/fs.go:123]: Thank you @timothyb89, this is very helpful.  Let me look at this closer

Comment (jentfoo) [reply lib/utils/fs.go]: That said, I am surprised to see an OS compatibility issue with `Lstat`: https://github.com/gravitational/teleport/actions/runs/6150975451

Should this be os conditional?

Comment (jentfoo) [reply lib/utils/fs.go]: I see the issue, working on a fix now

Comment (timothyb89) [reply lib/utils/fs.go]: I'm pretty sure that covers it. I've been feeding it inputs and haven't been able to break it, so that's something :slightly_smiling_face:

Comment (jentfoo) [reply lib/utils/fs.go]: Same!  I have been mentally weighing if I should build unit tests to cover these cases.  The downside is building the directory structure for testing is a fair bit of complexity.  Let me know your thoughts

File: lib/utils/fs.go
Comment (Tener) [inline lib/utils/fs.go:138]: This code will be somewhat painful to test, but I do think we should test it anyway due to its complexity.

**Comment by Tener** [comment]
Independently, these two options are meaningful in this context:
- https://docs.kernel.org/admin-guide/sysctl/fs.html#protected-hardlinks
- https://docs.kernel.org/admin-guide/sysctl/fs.html#protected-symlinks

These are available in Linux kernel 3.6+, released 11 years ago, and enabled in _most_ contexts (but not all).

I don't think such an option is available on MacOS, though. It surely isn't the default.

File: lib/utils/fs_test.go
Comment (Tener) [inline lib/utils/fs_test.go:40]: There is a handy utility, which handles cleanup too:
```suggestion
	rootDir := t.TempDir()
```

File: lib/utils/fs.go
Comment (Tener) [inline lib/utils/fs.go:138]: We may want to add the original file path in the error; see my comment in tests for example why this could be useful.
```suggestion
				return nil, trace.BadParameter("failed to open file %v, symlink not allowed in path: %v", path, subPath)
```

File: lib/utils/fs_test.go
Comment (Tener) [inline lib/utils/fs_test.go:75]: This one actually fails on my machine, because `/var` is symlink to `/private/var`:

```
=== RUN   TestOpenFileSymlinks/directFileOpenAllowed
    fs_test.go:73: 
        	Error Trace:	/Users/tener/code/teleport/lib/utils/fs_test.go:73
        	Error:      	Received unexpected error:
        	            	symlink not allowed in path: /var
        	Test:       	TestOpenFileSymlinks/directFileOpenAllowed
    --- FAIL: TestOpenFileSymlinks/directFileOpenAllowed (0.00s)
```

I wonder if we can make the error more informative, i.e. include the information about the file we were trying to access:
```
=== RUN   TestOpenFileSymlinks/directFileOpenAllowed
    fs_test.go:77: 
        	Error Trace:	/Users/tener/code/teleport/lib/utils/fs_test.go:77
        	Error:      	Received unexpected error:
        	            	failed to open file /var/folders/c0/_6x4ddps6zn5js_4y31xlz380000gn/T/TestOpenFileSymlinks2354108266/001/dir/file, symlink not allowed in path: /var
        	Test:       	TestOpenFileSymlinks/directFileOpenAllowed
    --- FAIL: TestOpenFileSymlinks/directFileOpenAllowed (0.00s)
```

See the suggestion above. For this particular test, I think this works:

```suggestion
	t.Run("directFileOpenAllowed", func(t *testing.T) {
		filePath, err = filepath.EvalSymlinks(filePath)
		require.NoError(t, err)
		f, err := openFile(filePath, false)
		require.NoError(t, err)
		defer f.Close()
	})
```

Alternatively, you may want to apply this right next to declaration of `filePath` variable:

```golang
	filePath := filepath.Join(dirPath, "file")
	_, err = os.Create(filePath)
	require.NoError(t, err)
	filePath, err = filepath.EvalSymlinks(filePath)
	require.NoError(t, err)
```

**Review by Tener — APPROVED** [review]

**Comment by Tener** [comment]
From the compatibility point of view, this is a breaking change. I don't think it will affect many configurations, but it is possible, in principle, that one of the paths in which we will now disallow symlinks contains those. For example, you can have `/home/%username%/` be a home dir for users, but symlinked to `/mnt/networkshare/developers/%username%`. For that reason, I'd add appropriate label as well as avoid backports to old releases.

File: lib/utils/fs.go
Comment (codingllama) [inline lib/utils/fs.go:108]: nits:

```suggestion
// OpenFileNoSymlinks opens a file, ensuring it's an actual file and not a directory or symlink.
```

File: lib/utils/fs.go
Comment (codingllama) [inline lib/utils/fs.go:103]: nit:

```suggestion
// OpenFileAllowingSymlinks opens a file, if it's a symlink the file referenced by the link will be associated to
// the returned os.File.  Returns an error if the file is not found or is a directory.
```

Comment (codingllama) [inline lib/utils/fs.go:103]: > if it's a symlink the file referenced by the link will be associated to
> // the returned os.File

What do you mean here? Could we say it resolves the links instead?

File: lib/utils/fs.go
Comment (codingllama) [inline lib/utils/fs.go:114]: I think we can kill this comment, the public methods should cover all of this.

```suggestion
```

File: lib/utils/fs.go
Comment (codingllama) [inline lib/utils/fs.go:134]: Is the Lstat skip in the first path component on purpose?

File: lib/utils/fs.go
Comment (codingllama) [inline lib/utils/fs.go:133]: Suggestion: Use a for-range, Join addictively, check first path component, spacing.

```suggestion
		components := strings.Split(newPath, string(os.PathSeparator))
		var subpath string
		for _, p := range components {
			subpath = filepath.Join(subpath, p)

			// The first component in an absolute path is empty.
			if subpath == "" {
				subPath = string(os.PathSeparator)
			}

```

File: lib/utils/fs.go
Comment (codingllama) [inline lib/utils/fs.go:138]: nit: Skip filler words like "failed" in error chains.

```suggestion
				return nil, trace.BadParameter("opening file %v, symlink not allowed in path: %v", path, subPath)
```

File: lib/utils/fs_test.go
Comment (codingllama) [inline lib/utils/fs_test.go:68]: Let's refactor this into a table-driven test, it's a good fit.

https://github.com/golang/go/wiki/TableDrivenTests

File: lib/utils/fs_test.go
Comment (codingllama) [inline lib/utils/fs_test.go:72]: Please test the public functions instead.

File: lib/utils/fs_test.go
Comment (codingllama) [inline lib/utils/fs_test.go:44]:
```suggestion
	dirFilePath := filepath.Join(dirPath, "file")
```

So the hierarchy is clearer from the names? Same for others.

File: lib/utils/fs_test.go
Comment (codingllama) [inline lib/utils/fs_test.go:83]: Maybe require.ErrorContains, so we can be sure we are hitting the expected condition?

```suggestion
		require.ErrorContains(t, err, "symlink")
```

Same for others.

(In a minor nit, we could be using `assert` instead of `require` here, which is the better habit.)

**Comment by codingllama** [comment]
> From the compatibility point of view, this is a breaking change

Could we mention in the changelog, for example, the places where symlinks are not allowed anymore? (Ie, the places affected by the change.)

@jentfoo

**Comment by jentfoo** [comment]
Compatibility options are being discussed here: https://github.com/gravitational/teleport-private/issues/1009#issuecomment-1716028879

I think we can find a way to backport this improvement while maintaining compatibility.

**Comment by codingllama** [comment]
> I think we can find a way to backport this improvement while maintaining compatibility.

I fail to see how the behavioral change would be non-breaking. I'm not saying to not backport, no strong opinions there on my part, just asking that we document whatever gets changed.

File: lib/utils/fs.go
Comment (jentfoo) [reply lib/utils/fs.go]: Wording updated in a216f6d14848719cd061eb29b1ed657247eaee25

File: lib/utils/fs.go
Comment (jentfoo) [reply lib/utils/fs.go]: Removed in a216f6d14848719cd061eb29b1ed657247eaee25

File: lib/utils/fs.go
Comment (jentfoo) [reply lib/utils/fs.go:136]: It's unclear to me why this is better, let me know your thoughts

File: lib/utils/fs.go
Comment (jentfoo) [reply lib/utils/fs.go:137]: I don't believe there is any skipping of components, can you explain further?

File: lib/utils/fs_test.go
Comment (jentfoo) [reply lib/utils/fs_test.go]: Now that the public functions include OS specific logic, I don't believe that would be a good fit for testing.

File: lib/utils/fs_test.go
Comment (jentfoo) [reply lib/utils/fs_test.go]: Updated with the other test updates in 734437505efc1fb43e6a9a478b2d2516a077085d

**Comment by jentfoo** [comment]
The compatibility solution was not as easy as it appeared.  For that reason I have removed the backport labels, but would like to discuss if we can get this into Teleport 14.

@codingllama and @Tener I would appreciate another review, thank you for all the feedback so far!

File: lib/utils/fs.go
Comment (codingllama) [reply lib/utils/fs.go:136]: The for starts with `i := 1`, so for relative paths we are not checking if the very first component is a symlink, unless I read it wrong?

File: lib/utils/fs.go
Comment (codingllama) [reply lib/utils/fs.go:135]: for-ranges are usually preferable as there is no need to deal with `i`. This ties in with a previous comment.

The snippet also replaces filepath.IsAbs with an empty string check, which I'd expect to be much cheaper - not a big deal here, but why not do it?

Other than those, it's a toss, hence the "suggestion" tag.

File: lib/utils/fs.go
Comment (jentfoo) [reply lib/utils/fs.go:136]: The components are selected using the sub-slice `[:i]`.  So the beginning part of the components will always be included in every iteration.  We start at `1` because the sub-slice is exclusive to the index.  This is also why the conditional is `<=`.

To me this seemed like the most obvious implementation but now I am wondering your recommendation to rebuild the path using the range of the components is easier to read.  Let me know your thoughts.

File: lib/utils/fs_test.go
Comment (codingllama) [reply lib/utils/fs_test.go]: If that's a concern, then we should be writing per-OS testing. I would still encourage you to test the public API and not internals.

File: lib/utils/fs.go
Comment (codingllama) [reply lib/utils/fs.go:136]: Gotcha, I did read it wrong. Thanks for explaining.

File: lib/utils/fs.go
Comment (jentfoo) [reply lib/utils/fs.go]: Updated, you convinced me this is more readable.  Thank you!

File: lib/utils/fs_test.go
Comment (codingllama) [inline lib/utils/fs_test.go:159]: nit: Close() should happen based on the actual outcome of the method, not be based on the test expectations, as the latter may break.

File: lib/utils/fs_test.go
Comment (jentfoo) [reply lib/utils/fs_test.go]: As in have the testing define the expected state depending on the OS it's run on?  Or have individual tests for each OS?

The first seems pretty reasonable if you would like me to update to that.  I will probably add it as additional testing rather than replace these.

Comment (codingllama) [reply lib/utils/fs_test.go]: I meant the latter - we have a vanilla test suite for the "regular" scenarios, but test macOS vs Linux behavior separately. We have a few packages that behave like that, for example webauthn tends to vary between darwin, linux and windows.

File: lib/utils/fs.go
Comment (codingllama) [reply lib/utils/fs.go]: Thanks!

**Review by codingllama — COMMENTED** [review]
Looks good, holding out just in case you do more test changes.

File: lib/utils/fs_test.go
Comment (jentfoo) [reply lib/utils/fs_test.go]: I am not sure I fully agree.  Because of the need to have a more complex setup that interacts with the filesystem I think it's best if this test is as centrally located as possible.

That said I did add testing around the public functions and the OS specific behaviors with hard links.  Let me know what you think.  Thank you!

**Review by codingllama — APPROVED** [review]

File: lib/utils/fs.go
Comment (r0mant) [inline lib/utils/fs.go:147]: This will not compile on Windows, but I don't remember off the top of my head if `tsh` (the only tool we build for Windows right now) pulls in this package. Might be worth checking before trying to merge this.

**Review by r0mant — APPROVED** [review]
I don't have objections to including this change in Teleport 14 (bar potential compilation issue) but @jentfoo can you please add a section describing the user impact of this change to our [draft release notes](https://docs.google.com/document/d/1NRaue1huhkzwmbDM6it8cNgaO9rZQ2b9O0w0nrG5W5Y/edit) under the "Breaking changes".

I also assume you did a smoke test with this change (e.g. starting teleport and starting a session)?

**Comment by zmb3** [comment]
We also build tctl for windows. We don't ship it yet, but we do build it on push and a lot of work went in to getting this far so please don't break it 🙂

File: lib/utils/fs.go
Comment (r0mant) [reply lib/utils/fs.go:147]: And yeah, `tctl` as well as @zmb3 mentioned below. Probably worth it to split this out into OS-specific files via build tags.

**Review by Tener — APPROVED** [review]

File: lib/utils/fs_unix.go
Comment (codingllama) [inline lib/utils/fs_unix.go:32]: nit: It's more usual to place the boolean return last:

```suggestion
func getHardLinkCount(fi os.FileInfo) (uint64, bool) {
```

In this case, I suppose returning just `0` or `-1` (without a bool) could work too. Up to you.

Comment (jentfoo) [reply lib/utils/fs_unix.go]: I changed the order.  The original order made sense to me based off how they should be evaluated, but no preference really.

I think it makes sense to return a bool for two reasons:
* I want to return a uint to match the `Stat_T` typing and avoid overflow risks
* Technically having a zero hardlink count is a valid condition if the file handle does not exist on disk (for example an open file handle that has been deleted)

Comment (codingllama) [reply lib/utils/fs_unix.go]: I'm happy with the changed order, thanks.

File: lib/utils/fs_unix.go
Comment (zmb3) [inline lib/utils/fs_unix.go]: I'd write this slightly differently in order to eliminate the else and align the happy path to the left.

```
statT, ok := fi.Sys().(*syscall.Stat_t)
if !ok {
    return false, 0
}
return true, statT.Nlink
```

File: lib/utils/fs_windows.go
Comment (zmb3) [inline lib/utils/fs_windows.go:28]: This comment used to apply to the entire file, not is only applies to `getPlatformLockFilePath`. I'd move it down.

File: lib/utils/fs_windows.go
Comment (zmb3) [inline lib/utils/fs_windows.go]: Maybe add a comment

```
// hard links not supported for Windows
```

File: lib/utils/fs_windows.go
Comment (jentfoo) [reply lib/utils/fs_windows.go]: Moved

File: lib/utils/fs_windows.go
Comment (jentfoo) [reply lib/utils/fs_windows.go]: Added a comment describing the current situation

File: lib/utils/fs_unix.go
Comment (jentfoo) [reply lib/utils/fs_unix.go]: This seems basically the same to me.  I tend to prefer avoiding `!` conditionals when logic for both states needs to be handled anyways.

**Comment by jentfoo** [comment]
@r0mant I have a suggestion in the release notes for how to describe this.  I tried to describe it fairly generically, but the only known use case to me is the loading of the `.tsh/environment` file.  Feel free to DM me if you think there is any better way to describe this for our customers.

Also I have added the `branch/v14` backport label so we can include this as part of Teleport 14

**Review by timothyb89 — APPROVED** [review]

**Pull request merged**
